zypper-migration

```diff
@@ -106,6 +106,43 @@ def check_if_param(opt, message)
   end
 end
 
+def products_restore()
+  # while running zypper-migration, one use case can lead to
+  # migrate from distribution [A] to [B] where A is different of B.
+  # This can lead to a scenario where, during that migration, a mismatch
+  # is produced due to this difference between distributions, making
+  # the migration to fail every time. To avoid that, there is an approach
+  # where the products information is backed up in tmp_products_path
+  # and restored if failed before rollingback in this plugin.
+  tmp_products_path = "/tmp/products.d.backup/"
```

Comment:
Fixed path in /tmp? I don't think that's a good idea and would survive a security review.

Reply:
I don't understand what this is supposed to do.
AFAIK products.d is changed only by installing a package.
If zypper fails after it already has installed this package, then the exit code must be 8 and zypper-migration fails on fs_inconsistent (line 596), prints the message about rollback and does not attempt to restore anything.

Reply:
True, but the zypper-migration code is being called inside a system migration.
In one of the steps of that migration, the base product gets updated (deleted).
Before doing so, the products data is saved as backup in case zypper-migration fails 
and it has to be restored.

Reply:
1. we backup and restore all relevant data, there is absolut no need to invent something new. If you think we need to backup and restore more data, add it to the relevant functions, but don't introduce something new, which is a pure security nightmare. Never ever use fixed paths in /tmp!

2. The product files get never deleted, they get updated via RPMs. If you are in the situation, that an update stops in the middle, means /etc/products.d does not contain the old RPMs anymore, restoring the files will not help you at all. The system is in an inconsistent state, with a mix of old and new RPMs. Either you have btrfs with snapshots, so you can do a rollback, or you have to do a fresh installation. Backup and restore of /etc/products.d will not help you at all to get your system into a consistent state again. For this, you would need to do a full backup and restore.

Reply:
A dumb question:
- Don't we have to clear products.d in order to get migration across major distributions versions?
  + At some point in the migration I know that the code queries the products, so if that says "SLES 12" for example there is no option to move to "SLES 15"
- To that effect there exist failure modes outside the migration plugin where a roll back is appropriate, so we should restore products.d content

- I only partially agree with the idea that if you end up in an inconsistent state one has to start from scratch. If we do fail for whatever reason and lets say some new "*-release" packages were installed but not all of the necessary release packages have been installed, then if products.d is restored to pre-migration conditions a re-registration of the system followed by "zypper dup" has a reasonably good chance of restoring the currently inconsistent state to the previous consistent state. I do think there is value in backing up the content of products.d and restoring it in certain conditions.

Reply:
No, there is absolute no need to clear products.d in order to get migration across major distributions if you use the correct SCC/SMT/RMT API.
Quite the opposite, clearing this directory could have bad sideeffects or prevent you from a migration, as required informations for SCC/SMT/RMT to calculate the migration target could be missing.

Reply:
@thkukuk @nadvornik something doesn't appear to fit together. In our testing and @jesusbv has more details migration across major distributions fails if `<target>` which is a sub-element of `<register>` exists. It appears that as long as the product target exists we cannot get across the SLE-12 to SLE-15 hurdle.

It was pointed out that

> if you use the correct SCC/SMT/RMT API.

but for the migration system we are not directly consuming any of the SCC/SMT/RMT APIs and we rely completely on the existing implementation. Thus this comment would indicate that the existing implementation does not use the correct APIs. I am not sufficiently familiar with the implementation of migration nor the SCC/SMT/RMT apis to form an opinion, we can only report the symptom. And the symptom is that we observe failures when product definitions exist and the `target` is set.

Reply:
Before implementing any workarounds, can you please report a bug on SCC?

Reply:
Which command is failing, at which point in time, and what are you doing before?
I cannot reproduce it, plain "zypper migration -p SLES/15/x86_64 --root <path>" from SLES12 SP4 to SLES15 is working fine with me. There must be something wrong in your environment.

Reply:
@thkukuk the command it runs is  ```zypper --root /system-root --releasever 15 ref -f```
here it is a extracted info from the log after running the migration:

```bash
Executing 'zypper --root /system-root --non-interactive --gpg-auto-import-keys  refresh'
Refreshing service 'cloud_update'.
...
All repositories have been refreshed.
Upgrading products
...
Executing 'zypper --root /system-root --releasever 15 ref -f'

Warning: Enforced setting: $releasever=15
Refreshing service 'cloud_update'.
Warning: There are no enabled repositories defined.
Use 'zypper addrepo' or 'zypper modifyrepo' commands to add or enable repositories.
Executing 'zypper --root /system-root --releasever 15 ref -f': RuntimeError: Refresh of repositories failed.

Migration failed.
```
Hope it helps

Reply:
This seems to be a small part of the output, and the important parts are missing. "Warning: There are no enabled repositories defined." means you did no migration before you try to refresh the repositories.

Reply:
yes, I did no migration before that command runs but, the command that triggers all that info is
```bash
zypper migration --non-interactive --gpg-auto-import-keys --no-selfupdate --auto-agree-with-licenses --strict-errors-dist-migration --replacefiles --product SLES/15/x86_64 --root /system-root
```
Then the info is as shown above.